### PR TITLE
GDV-28: [C++] Add hash functions on all data types

### PR DESCRIPTION
**Pull request by pravindra**

### CODE REVIEW

File: ci/travis_before_script_cpp.sh [hunk 1]
```diff
@@ -75,14 +75,19 @@ else
 fi
 
 # Do the style checks
-$TRAVIS_MAKE stylecheck
+$TRAVIS_MAKE stylefix
+if ! git diff-index --quiet HEAD --; then
+  # The syle fix required some changes
+  echho "Stylecheck failed, requires the following fix"
```
Comment (vvellanki) [inline ci/travis_before_script_cpp.sh]: Typo - should be echo
Should we mention that stylefix will automatically fix the issues flagged?

File: cpp/integ/hash_test.cc [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  auto literal_10 = TreeExprBuilder::MakeLiteral((int32_t)10);
+  auto hash32 = TreeExprBuilder::MakeFunction("hash32", {node_a, literal_10}, int32());
+  auto hash64 = TreeExprBuilder::MakeFunction("hash64", {node_a}, int64());
+  auto expr_0 = TreeExprBuilder::MakeExpression(hash32, res_0);
+  auto expr_1 = TreeExprBuilder::MakeExpression(hash64, res_1);
+
+  // Build a projector for the expression.
+  std::shared_ptr<Projector> projector;
+  Status status = Projector::Make(schema, {expr_0, expr_1}, pool_, &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
```
Comment (vvellanki) [inline cpp/integ/hash_test.cc]: Why do we need the "<<" operator?

File: cpp/src/precompiled/hash.cc [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+#include "./types.h"
+
+static inline uint64 rotate_left(uint64 val, int distance) {
+  return (val << distance) | (val >> (64 - distance));
+}
+
+//
+// MurmurHash3 was written by Austin Appleby, and is placed in the public
+// domain.
+// See http://smhasher.googlecode.com/svn/trunk/MurmurHash3.cpp
```
Comment (praveenbingo) [inline cpp/src/precompiled/hash.cc]: does dremio use murmur too?do we ensure that the two codes are same ?

File: cpp/src/precompiled/hash.cc [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+FORCE_INLINE int32 hash32(double val, int32 seed) {
+  return (int32)murmur3_64(double_to_long_bits(val), seed);
+}
+
+// Wrappers for all the numeric/data/time arrow types
+
+#define HASH64_WITH_SEED_OP(NAME, TYPE)                                              \
+  FORCE_INLINE                                                                       \
+  int64 NAME##_##TYPE(TYPE in, boolean is_valid, int64 seed, boolean seed_isvalid) { \
```
Comment (vvellanki) [inline cpp/src/precompiled/hash.cc]: Shouldn't there be a check for seed_isvalid?

File: ci/travis_before_script_cpp.sh (hunk 1 above)
Comment (pravindra) [reply ci/travis_before_script_cpp.sh]: yeah, travis failed because of this. still fixing.

File: cpp/integ/hash_test.cc (hunk 2 above)
Comment (pravindra) [reply cpp/integ/hash_test.cc]: it prints the error msg when the expect fails. useful for debugging.

File: cpp/src/precompiled/hash.cc (hunk 3 above)
Comment (pravindra) [reply cpp/src/precompiled/hash.cc]: yes, dremio uses murmur too.

they don't need to be strictly the same but it's useful for our testing if they are the same.  will add dremio tests for this.

File: cpp/src/precompiled/hash.cc (hunk 4 above)
Comment (pravindra) [reply cpp/src/precompiled/hash.cc]: the seed is actually a literal. but, our function type (NULL_NEVER/NULL_INTERNAL) require the is_valid bit to be passed for every param. 

checking for seed_isvalid is harmless, so I'll add it.

**Pull request merged**